### PR TITLE
fix: Fix Azure provider and add complex e2e testing

### PR DESCRIPTION
**Pull request by mattzh72**
## Description

Fix Azure endpoint and add complex e2e testing. Originally, Azure endpoint was broken. We fixed this by factoring out the logic written for the OpenAI provider and extending it the Azure provider.

Variety of other changes as well:
- Move configs to `tests` directory to make clear configs are only used for testing
- Refactor several files and logic around `inner_thoughts_in_kwargs`
- Rename some variables for coding style / maintainability
- Create the Azure provider class

## Testing

Wrote testing suite in `test_endpoints.py` and manually ran them. Holding off on adding this to CI for now because we need @cpacker to add environment variables into GH Actions.

### CODE REVIEW

File: tests/helpers/endpoints_helper.py [hunk 1]
```diff
@@ -77,13 +83,16 @@ def setup_agent(
     return agent_state
 
 
+# def
```
Comment (sarahwooders) [inline tests/helpers/endpoints_helper.py:86]: remove?

File: letta/schemas/llm_config.py [hunk 2]
```diff
@@ -34,7 +34,10 @@ class LLMConfig(BaseModel):
         "vllm",
         "hugging-face",
     ] = Field(..., description="The endpoint type for the model.")
-    model_endpoint: str = Field(..., description="The endpoint for the model.")
+    model_endpoint: Optional[str] = Field(None, description="The endpoint for the model.")
+    model_version: Optional[str] = Field(
```
Comment (sarahwooders) [inline letta/schemas/llm_config.py:38]: maybe rename this `api_version`?

File: tests/helpers/endpoints_helper.py (hunk 1 above)
Comment (mattzh72) [reply tests/helpers/endpoints_helper.py]: ack

File: letta/schemas/llm_config.py (hunk 2 above)
Comment (mattzh72) [reply letta/schemas/llm_config.py]: done

File: letta/credentials.py [hunk 3]
```diff
@@ -30,14 +30,14 @@ class LettaCredentials:
...
-    azure_key: Optional[str] = None
+    azure_key: Optional[str] = os.getenv("AZURE_OPENAI_API_KEY")
 
     # groq config
     groq_key: Optional[str] = os.getenv("GROQ_API_KEY")
 
     # base llm / model
-    azure_version: Optional[str] = None
-    azure_endpoint: Optional[str] = None
+    azure_version: Optional[str] = "2023-03-15-preview"  # None
```
Comment (sarahwooders) [inline letta/credentials.py:39]: you probably want to remove this defaults, since it can make debugging confusing

Comment (mattzh72) [reply letta/credentials.py:39]: SORRY i should've caught this, this is debugging litter from when i started looking at azure last week :\

**Review by sarahwooders — APPROVED** [review]

**Pull request merged**